src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,602 @@
+
+(**
+ * Specialization of the above to handle demotion while carrying 
+ * information on whether we have entered a for loop.
+ * (TODO: We can probably coerce this and the 
+ *  other query demotable function together)
+ * 
+ *)
+let query_demotable_funkinds_loop query_expr_loop in_loop kind exprs =
+  let ret_demotable_types =
+    Set.Poly.union_list (List.map ~f:query_all_expr_eigen_names exprs)
+  in
+  match kind with
+  | Fun_kind.StanLib (name, (_ : bool Fun_kind.suffix), _) -> (
+    match name with
+    | x
+      when Stan_math_signatures.is_reduce_sum_fn x
+           || Stan_math_signatures.is_variadic_ode_fn x ->
+        ret_demotable_types
+    | _ ->
+        let is_fun_support =
+          Stan_math_signatures.query_stan_math_mem_pattern_support name
+            (List.map ~f:find_args exprs)
+        in
+        if is_fun_support then
+          Set.Poly.union_list (List.map ~f:(query_expr_loop in_loop) exprs)
+        else ret_demotable_types )
+  | CompilerInternal (_ : Internal_fun.t) -> Set.Poly.empty
+  | UserDefined ((_ : string), (_ : bool Fun_kind.suffix)) ->
+      ret_demotable_types
+
+(**
+ * Find indices on Matrix and Vector types that perform single 
+ *  cell access. Returns true if it finds
+ * a vector, row vector, matrix, or matrix with single cell access
+ * as well as an array of any of the above that is accessing the 
+ * inner matrix types cell.
+ *)
+let rec is_uni_eigen_loop_indexing (ut : UnsizedType.t)
+    (index : 'a Expr.Fixed.t Index.t list) =
+  match (ut, index) with
+  | (UnsizedType.UArray t | UFun (_, ReturnType t, _, _)), index -> (
+    match List.tl index with
+    | Some cut_list -> is_uni_eigen_loop_indexing t cut_list
+    | None -> is_uni_eigen_loop_indexing t [Index.All] )
+  | (UVector | URowVector), [Index.Single (_ : 'a Expr.Fixed.t)] -> true
+  | UMatrix, [Single (_ : 'a Expr.Fixed.t); Single (_ : 'a Expr.Fixed.t)] ->
+      true
+  (* None of the below contain single cell access*)
+  | (UReal | UInt | UMathLibraryFunction | UFun (_, Void, _, _)), _ -> false
+  | (UVector | URowVector), _ -> false
+  | UMatrix, _ -> false
+
+(**
+ * Specialization to find demotable (SoA -> AoS) objects
+ *  that carries logic on whether we are in a for loop or not.
+ * (TODO: Could be combined with other query demotable above.)
+ *)
+let rec query_demotable_exprs_loop in_loop Expr.Fixed.({pattern; _}) =
+  let query_expr = query_demotable_exprs_loop in_loop in
+  match pattern with
+  | FunApp (kind, (exprs : Expr.Typed.Meta.t Expr.Fixed.t list)) ->
+      query_demotable_funkinds_loop query_demotable_exprs_loop in_loop kind
+        exprs
+  | Var (_ : string) | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) ->
+      Set.Poly.empty
+  | TernaryIf (predicate, texpr, fexpr) ->
+      Set.Poly.union
+        (Set.Poly.union (query_expr predicate) (query_expr texpr))
+        (query_expr fexpr)
+  | Indexed (Expr.Fixed.({pattern= Var name; meta= {type_; _}}), indexed) ->
```

Comment:
Is it enough to only check variables? What about something like `transpose(matrix)[5,5]`?

Reply:
Oh good catch! I'll check this out

Reply:
Hmm, so I fixed this up, but thinking about it more I'm not sure which heuristic is best. So like if someone does like 
`exp(x)[1]`, then I'd actually argue the `exp(x)` is going to dominate the time. Though `transpose(matrix)[5,5]` is a case where where `transpose()` will probably not dominate the time. @rok-cesnovar @t4c1 do yinz have thoughts about this? Maybe the general rule actually makes since to only ban SoA when doing direct subsetting on a variable, but if we see a pattern like the above (which is odd but can happen) then leave it as SoA?

Reply:
For context, we are talking about single cell access and assignment in for loops and whether objects used in statements such as `x_in_question` in the below should be SoA or AoS.

```stan
for (i in 1:N) {
  target += transpose(x_in_question)[5, 5];
}
```

Looking at the above I'm inclined to think `x_in_question` should be demoted to `AoS`. Though  I'm kind of 50/50

Reply:
My general heuristic is we should only use `SoA` when we know 100% it's a good idea. This is a wishy washy area so I'd rather be conservative

Reply:
If the expression before indexing is evaluated, both options would be quite slow compared to what would be the optimal thing to compute - moving indexing to the arguments of the expression instead of the result. I think it would be good to do that in some optimization pass regardless of AoS/SoA. How that is done depends on the functions used in the expression. Many functions simply let us use the same index for arguments, but there are exceptions (transpose, matrix multiply ...)

However, we do not evaluate expressions before indexing. That is more performant when used with AoS, as Eigen makes sure only the indexed part of the expression is computed. The twist here is that not all Eigen expressions support indexing. I bet the following would fail to compile: 

```
matrix a,b;
real c = (a * b)[0,0];
```

Reply:
> If the expression before indexing is evaluated, both options would be quite slow compared to what would be the optimal thing to compute - moving indexing to the arguments of the expression instead of the result.

Yeah it would be nice if people didn't write this, but we don't really get to decide that for users.

> I think it would be good to do that in some optimization pass regardless of AoS/SoA. How that is done depends on the functions used in the expression. Many functions simply let us use the same index for arguments, but there are exceptions (transpose, matrix multiply ...)

Yes there is an optimization to do that, though it's only turned on if all other optims are turned on. I think we want more testing for it.

> However, we do not evaluate expressions before indexing. That is more performant when used with AoS, as Eigen makes sure only the indexed part of the expression is computed. The twist here is that not all Eigen expressions support indexing. I bet the following would fail to compile:

It feels like we should have a `to_ref()` inside of the rvalue single indexing then else people are going to get the wrong answer when calling code like the above.

But eod it sounds like we should just use AoS here so the logic I updated this with should work